### ISSUE TITLE
Prod Mode Compiling Out TypeScript decorator metadata from external project

### ISSUE THREAD

**Issue by amcdnl** [angular/angular-cli#10584]
### Observed behavior
The cli in prod mode is compiling out the typescript decorator (custom decorator) metadata from a external project compiled using ng-packagr.

An example of the decorator metadata would be:

```TS
__decorate([
    Action(Navigate),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Navigate]),
    __metadata("design:returntype", void 0)
], RouterState.prototype, "navigate", null);
```

The result can be seen in this screenshot on the left:

![img](https://content.screencast.com/users/amcdaniel22/folders/Snagit/media/19efddb3-b7a2-4295-80bd-285fdadeac11/2018-05-01_16-56-40.png)

It seems to be related to only external projects. I'm using the same custom decorator used in this code in my project and it works fine.

### Desired behavior
It not to compile this out.

### Repro steps
- `npm i @ngxs/router-plugin @ngxs/store`
- Include the store and router plugin in your project like so:

```TS
@NgModule({
  imports: [
    NgxsModule.forRoot([]),
    NgxsRouterPluginModule.forRoot()
  ],
  declarations: [AppComponent],
  bootstrap: [AppComponent]
})
export class AppModule {}
```

- Build the application in production
- Inspect the source and note the decorate metadata was compiled out

For reference here is the [compiled code](https://gist.github.com/amcdnl/0236f4849e2941b279113a5589cef44b#file-compiled-js-L207-L230)

### Versions
Node 10
Npm 6
Mac
"@angular/cli": "^6.0.0-rc.8",

**Comment by filipesilva** [comment]
@amcdnl does this repro fail at runtime with the changes you listed? If not, can you give me a repro that fails so I can see something failing and then try to fix it?

**Comment by deebloo** [comment]
What is worth noting is that the router plugin mentioned here uses ngxs internally and is then packaged with ng-packagr

**Comment by filipesilva** [comment]
I looked into this a bit and it doesn't seem to have anything to do with production mode. 

Using this `src/app/app.component.ts`:
```
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { NgxsRouterPluginModule } from '@ngxs/router-plugin';
import { NgxsModule } from '@ngxs/store';

import { AppComponent } from './app.component';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    NgxsModule.forRoot([]),
    NgxsRouterPluginModule.forRoot()
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

And Angular CLI 6.0.0-rc.8, `ng serve --aot` will show the following message at runtime:
```
Uncaught Error: StaticInjectorError(AppModule)[RouterState -> Router]: 
  StaticInjectorError(Platform: core)[RouterState -> Router]: 
    NullInjectorError: No provider for Router!
    at NullInjector.push../node_modules/@angular/core/fesm5/core.js.NullInjector.get (core.js:979)
    at resolveToken (core.js:1232)
    at tryResolveToken (core.js:1182)
    at StaticInjector.push../node_modules/@angular/core/fesm5/core.js.StaticInjector.get (core.js:1077)
    at resolveToken (core.js:1232)
    at tryResolveToken (core.js:1182)
    at StaticInjector.push../node_modules/@angular/core/fesm5/core.js.StaticInjector.get (core.js:1077)
    at resolveNgModuleDep (core.js:9235)
    at _createClass (core.js:9276)
    at _createProviderInstance$1 (core.js:9252)
```

I'm not sure why this happens but it doesn't seem like a Uglify or Build Optimizer problem. It seems related to the issues in this PR https://github.com/angular/angular-cli/pull/10060.

**Comment by filipesilva** [comment]
Btw if this isn't the same error as you are seeing, please provide a reproduction that I can clone and shows the right one.

**Comment by deebloo** [comment]
Can you try adding the router module and see if that gets you to the wrror?

**Comment by deebloo** [comment]
*error

**Comment by amcdnl** [comment]
@filipesilva = whops, I forgot to tell you to include the angular router. There is no runtime error, it compiles out the code so nothing actually runs on those decorators.

**Comment by amcdnl** [comment]
I also fixed an issue where my TypeScript enums were getting compiled out too. I had a enum like:

```TS
export enum Foo {
  Bar = 'BAR'
}
```

and it was resolving to `undefined`. I rewrote it to be:

```TS
export const enum Foo {
   Bar = 'BAR'
}
```

and it worked then. I wonder if that is related to this too.

**Comment by filipesilva** [comment]
@amcdnl I have used your repro and can confirm I see the problem. @clydin and I have analysed what's happening and can explain what's going wrong.

These libraries have `"sideEffects": false` in their `package.json`, which prompts advanced optimizations, but they actually have side effects.

If you look at your gist: https://gist.github.com/amcdnl/0236f4849e2941b279113a5589cef44b#file-compiled-js-L207-L230:

```
console.log('!!!!!!!');
__decorate([
    Action(Navigate),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Navigate]),
    __metadata("design:returntype", void 0)
], RouterState.prototype, "navigate", null);
```

The toplevel `console.log` is a side effect (which will be removed), and so is any `__decorate` call.

I am not sure if you added `"sideEffects": false` yourself, or if it was added automatically by `ng-packagr`, but at the moment it should be removed. 

If you remove it directly in your `node_modules` you will also need to delete `node_modules/.cache` and `node_modules/@angular-devkit/build-optimizer/src/.cache` since these will cache code for production builds.
